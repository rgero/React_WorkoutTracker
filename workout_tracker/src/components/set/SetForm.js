import React, {useState} from 'react';

import '../../styles/components/inputs.css'

export const SetForm = ({set = {}, onSubmit})=> 
{
    const [reps, setReps] = useState(set.reps ? set.reps : "")
    const [weight, setWeight] = useState(set.weight ? set.weight : "");
    const [error, setError] = useState("")

    const processChange = (event, callback) => {
        const amount = event.target.value;
        if (!amount || amount.match(/^\d{1,}(\.\d{0,2})?$/)) {
            callback(amount);
        }
    }

    const processNewSet = (e) => {
        e.preventDefault();
        
        if (reps === "" || weight === "")
        {
            setError("Reps and Weight need to be populated");
            return;
        }

        let newSet = { reps, weight}
        setReps("");
        setWeight("");
        setError("");
        onSubmit(newSet);
    }

    return (
        <form className="form" onSubmit={processNewSet}>
            {error !== '' && <p className="form__error">{error}</p>}

            {/* Reps */}
            <div className="input-entry">
                <label>Reps</label>
                <input
                    type="text-input"
                    className="text-input"
                    placeholder="Reps"
                    autoFocus
                    value={reps}
                    onChange={e => processChange(e, setReps)}
                />
            </div>

            <div className="input-entry">
                <label>Weight</label>
                <input
                    type="text-input"
                    className="text-input"
                    placeholder="Weight"
                    value={weight}
                    onChange={e => processChange(e, setWeight)}
                />
            </div>

            <div>
                <button className="button">Save Set</button>
            </div>

        </form>
    )
}