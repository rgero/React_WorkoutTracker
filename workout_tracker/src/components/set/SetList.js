import React from 'react';

import ListGroup from 'react-bootstrap/ListGroup';
import Container from 'react-bootstrap/Container';
import Col from 'react-bootstrap/Col';
import Row from 'react-bootstrap/Row';

export const SetList = ({setList = [], onDelete=null})=> 
{
  const processDelete = (index) => {
    if (!onDelete)
    {
      return;
    }
    onDelete(index);
  }

  return (
    <Container>
      {
        setList.length === 0 ? (
          <div className="list-item list-item--message">
            <span>No Sets</span>
          </div>
        ) : (
          <Container>
            <ListGroup>
              <ListGroup.Item>
                <Row>
                  <Col>
                    Reps
                  </Col>
                  <Col>
                    Weight
                  </Col>
                </Row>
              </ListGroup.Item>
              {
                setList.map((set, index) => (
                  <ListGroup.Item key={`${set.reps}_${set.weight}_${index}`} onClick={(e) => processDelete(index)}>
                    <Row>
                      <Col>
                        {set.reps}
                      </Col>
                      { set.weight ? (
                        <Col>
                          {set.weight}
                        </Col>
                      ) : ( null ) }
                    </Row>
                  </ListGroup.Item>
                ))
              }
            </ListGroup>
          </Container>
        )
      }
    </Container>
  )
}