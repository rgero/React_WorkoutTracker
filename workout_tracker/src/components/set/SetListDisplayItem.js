import React from 'react';

export const SetListDisplayItem = ({set})=> 
{
    return (
        <div className="setListItem">
            <div>
                {set.reps}
            </div>
            <div>
                {set.weight}
            </div>
        </div>
    )
}