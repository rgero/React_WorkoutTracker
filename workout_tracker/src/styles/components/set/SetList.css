.setListItem {
    display: flex;
    flex-direction: row;
    justify-content: space-evenly;
}

.setListHeader
{
    display: flex;
    flex-direction: row;
    justify-content: space-evenly;
    font-weight: bold;
}

.setList
{
    margin: 1.2rem 0 1.2rem 0;
    border: 1px black solid;
    border-radius: 10px;
}