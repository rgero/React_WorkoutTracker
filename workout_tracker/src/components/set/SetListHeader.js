import React from 'react';

import '../../styles/components/set/SetList.css';

export const SetListHeader = ({reps, weight})=> 
{
    return (
        <div className="setListHeader">
            <div>
                Reps
            </div>
            <div>
                Weight
            </div>
        </div>
    )
}