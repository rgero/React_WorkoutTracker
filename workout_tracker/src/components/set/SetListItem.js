import React, {useState} from 'react';

import '../../styles/components/set/SetList.css';

export const SetListItem = ({set, onDelete})=> 
{
    const [color, setColor] = useState({background: "#ffffff"});

    const highlightColor = {background: "#03a9fc"}
    const unhighlightColor = {background: "#ffffff"}

    const onMouseOver = (e) => {
        setColor(highlightColor);
    }
    
    const onMouseOut = (e) => {
        setColor(unhighlightColor);
    }

    const itemClicked = (e) => {
        onDelete(set.id);
    }

    return (
        <div className="setListItem" style={color} onMouseEnter={onMouseOver.bind(this)} onMouseLeave={onMouseOut.bind(this)} onClick={itemClicked}>
            <div>
                {set.reps}
            </div>
            <div>
                {set.weight}
            </div>
        </div>
    )
}