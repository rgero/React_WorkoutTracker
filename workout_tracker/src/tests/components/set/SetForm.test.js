import TestRenderer from 'react-test-renderer';

import {render, screen} from '@testing-library/react';
import '@testing-library/jest-dom';
import userEvent from '@testing-library/user-event'

import {SetForm} from '../../../components/set/SetForm';
import { TestSet } from '../../testData';

describe("SetForm Rendering Tests", () => {
    test('No data', ()=> {
        let element = new TestRenderer.create(
            <SetForm onSubmit={jest.fn()} />
        )
        expect(element.toJSON()).toMatchSnapshot();
    })

    test('With Data', ()=> {
        let element = new TestRenderer.create(
            <SetForm
                set = {TestSet}
                onSubmit={jest.fn()} 
            />
        )
        expect(element.toJSON()).toMatchSnapshot();
    })
})

describe("SetForm Submissions", ()=> {
    let user;
    let submit;
    beforeEach(()=> {
        user = userEvent.setup();
        submit = jest.fn();

        render(<SetForm onSubmit={submit} />)
    })

    test("Valid Submission - Both values", async ()=> {
        let testSet = TestSet;
        const repInput = screen.getByRole("textbox", {name: /reps/i});
        const weightInput = screen.getByRole("textbox", {name: /weight/i});

        await user.click(repInput);
        await user.keyboard(testSet.reps);
    
        await user.click(weightInput);
        await user.keyboard(testSet.weight);
    
        const button = screen.getByRole("button");
        await user.click(button);
        

        expect(submit).toHaveBeenCalled();
        expect(submit).toHaveBeenCalledWith( testSet )
    })

    test("Valid Submission - Just Reps", async ()=> {
        let testSet = {
            reps: "20",
            weight: ""
        }

        const repInput = screen.getByRole("textbox", {name: /reps/i});

        await user.click(repInput);
        await user.keyboard(testSet.reps);
    
        const button = screen.getByRole("button");
        await user.click(button);
        
        expect(submit).toHaveBeenCalled();
        expect(submit).toHaveBeenCalledWith( testSet )
    })

    test("Invalid Submission - No Reps", async () => {
        let testSet = TestSet;

        const weightInput = screen.getByRole("textbox", {name: /weight/i});
      
        await user.click(weightInput);
        await user.keyboard(testSet.weight);

        const button = screen.getByRole("button");
        await user.click(button);

        expect(submit).not.toHaveBeenCalled();

        const errorField = screen.getByText("Error: Missing Reps");
        expect(errorField).toBeInTheDocument();
    })
})