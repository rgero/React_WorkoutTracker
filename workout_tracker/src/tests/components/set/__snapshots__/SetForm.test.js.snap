// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SetForm Rendering Tests No data 1`] = `
<div
  className="container-md"
>
  <form
    className=""
    onSubmit={[Function]}
  >
    <div
      className="justify-content-md-center align-items-md-center row row-cols-md-4"
    >
      <div
        className="col-md"
      >
        <div
          className="mb-3 form-floating"
        >
          <input
            aria-label="reps"
            className="form-control"
            id="floatingReps"
            onChange={[Function]}
            type="text"
            value=""
          />
          <label
            htmlFor="floatingReps"
          >
            Reps
          </label>
        </div>
      </div>
      <div
        className="col-md"
      >
        <div
          className="mb-3 form-floating"
        >
          <input
            aria-label="weight"
            className="form-control"
            id="floatingWeight"
            onChange={[Function]}
            type="text"
            value=""
          />
          <label
            htmlFor="floatingWeight"
          >
            Weight
          </label>
        </div>
      </div>
    </div>
    <button
      aria-label="setSubmit"
      className="float-end btn btn-outline-secondary"
      disabled={false}
      type="submit"
    >
      Add Set
    </button>
  </form>
</div>
`;

exports[`SetForm Rendering Tests With Data 1`] = `
<div
  className="container-md"
>
  <form
    className=""
    onSubmit={[Function]}
  >
    <div
      className="justify-content-md-center align-items-md-center row row-cols-md-4"
    >
      <div
        className="col-md"
      >
        <div
          className="mb-3 form-floating"
        >
          <input
            aria-label="reps"
            className="form-control"
            id="floatingReps"
            onChange={[Function]}
            type="text"
            value="20"
          />
          <label
            htmlFor="floatingReps"
          >
            Reps
          </label>
        </div>
      </div>
      <div
        className="col-md"
      >
        <div
          className="mb-3 form-floating"
        >
          <input
            aria-label="weight"
            className="form-control"
            id="floatingWeight"
            onChange={[Function]}
            type="text"
            value="200"
          />
          <label
            htmlFor="floatingWeight"
          >
            Weight
          </label>
        </div>
      </div>
    </div>
    <button
      aria-label="setSubmit"
      className="float-end btn btn-outline-secondary"
      disabled={false}
      type="submit"
    >
      Add Set
    </button>
  </form>
</div>
`;
